### PR TITLE
ossl_bsearch(): New generic internal binary search utility function

### PR DESCRIPTION
**Pull request by levitte**
OBJ_bsearch_ and OBJ_bsearch_ex_ are generic functions that don't
really belong with the OBJ API, but should rather be generic utility
functions.  The ending underscore indicates that they are considered
internal, even though they are declared publicly.

Since crypto/stack/stack.c uses OBJ_bsearch_ex_, the stack API ends up
depending on the OBJ API, which is unnecessary, and carries along
other dependencies.

Therefor, a generic internal function is created, ossl_bsearch().
This removes the unecessary dependencies.

##### Checklist

- [ ] documentation is added or updated

### CODE REVIEW

**Comment by levitte** [comment]
The (re?)discovery that lead to this was that the stack functionality depends on the OBJ API, which in turn depends on the ASN1 API.  So basically, anything that uses STACK ends up sucking in (parts of) ASN1.  It's important to break that chain, especially for the FIPS module.  That makes this related to #8728.

File: crypto/bsearch.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        p = &(base_[i * size]);
+        c = (*cmp) (key, p);
+        if (c < 0)
+            h = i;
+        else if (c > 0)
+            l = i + 1;
+        else
+            break;
+    }
+    if (c != 0 && !(flags & OSSL_BSEARCH_VALUE_ON_NOMATCH))
```
Comment (paulidale) [inline crypto/bsearch.c]: Wondering if ditching the flags argument and just making it a Boolean would be sensible?
It is effectively an overly complicated Boolean parameter.

**Comment by mattcaswell** [comment]
Looks good but Travis failures appear relevant.

**Comment by richsalz** [comment]
Do we use trailing underscore anywhere else?  Can we not create a new convention?

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

5c3f1e34b559c9b4372bf48aab63b61a6cd5edbb ossl_bsearch(): New generic internal binary search utility function

File: crypto/bsearch.c (hunk 1 above)
Comment (levitte) [reply crypto/bsearch.c]: That requires more investigation.  I think that's for another PR

Comment (paulidale) [reply crypto/bsearch.c]: Okay.

**Pull request closed without merging**